wordpress-mobile/WordPress-FluxC-Android
# Feature/mark referrers as spam

Partially solves [wordpress-mobile/WordPress-Android#11818](https://github.com/wordpress-mobile/WordPress-Android/issues/11818)

### Description

It adds the possibility to report and unreport a referrer as spam through the `ReferrersStore.kt` by implementing [/sites/$site/stats/referrers/spam/new](https://developer.wordpress.com/docs/api/1.1/post/sites/%24site/stats/referrers/spam/new/) and [/sites/$site/stats/referrers/spam/delete](https://developer.wordpress.com/docs/api/1.1/post/sites/%24site/stats/referrers/spam/delete/) endpoints.

### Changes

- Added the new endpoints in `wp-com-endpoints.txt`
- Added `reportReferrerAsSpam()` and `unreportReferrerAsSpam()` suspend methods in `ReferrersRestClient.kt`
- Added `reportReferrerAsSpam()` and `unreportReferrerAsSpam()` suspend methods in `ReferrersStore.kt`
- Added the error `StatsErrorType.ALREADY_SPAMMED` so apps can know if a referrer was marked as spam.
- Added unit tests in `ReferrersRestClientTest.kt` and `ReferrersStoreTest.kt`
- Added an integration in `ReleaseStack_TimeStatsTestJetpack.kt`
- Renamed some methods and classes that were called `ReferrersResponse` to `FetchReferrersResponse`
- `reportReferrerAsSpam()` and `unreportReferrerAsSpam()` updates the cache marking the referrer as spam

### To test

Run `testReportReferrerAsSpam()` in `ReleaseStack_TimeStatsTestJetpack.kt`. This will:

- Fetch the list of referrers of the first site retrieved from the `BuildConfig.TEST_WPCOM_USERNAME_SINGLE_JETPACK_ONLY` credentials
- Unreport the first referrer as spam and assert the response was successful 
- Report the first referrer as spam and assert the response was successful. Do it again and assert that the response returns an error of `StatsErrorType.ALREADY_SPAMMED`

Please note that the test will fail if the first site doesn't have any referrers.
